### PR TITLE
 🐛 Fix incorrect handling when removing from the end of the triggers array

### LINKED ISSUE #1699: Removing triggers from ScaledObject doesn't reflect in the HPA it manages
### Report

I have a ScaledObject with several rabbitmq triggers, and when deleting one of them from the SO definition, changes are not propagated to the underlying HPA. Adding triggers work fine though.

Example SO:
```yaml
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: consumer
  namespace: keda2
spec:
  advanced:
    horizontalPodAutoscalerConfig:
      behavior:
        scaleDown:
          stabilizationWindowSeconds: 60
        scaleUp:
          stabilizationWindowSeconds: 60
  scaleTargetRef:
    name: consumer
  pollingInterval: 5
  minReplicaCount: 1
  maxReplicaCount: 10
  triggers:
  - type: rabbitmq
    metadata:
      protocol: amqp
      queueName: keda-tests
      queueLength: "5"
    authenticationRef:
      name: keda-trigger-auth-rabbitmq-conn
  - type: rabbitmq
    metadata:
      protocol: amqp
      queueName: keda-tests2
      queueLength: "5"
    authenticationRef:
      name: keda-trigger-auth-rabbitmq-conn
```

which creates the following HPA:
```bash
kubectl get hpa keda-hpa-consumer
NAME                REFERENCE             TARGETS                MINPODS   MAXPODS   REPLICAS   AGE
keda-hpa-consumer   Deployment/consumer   0/5 (avg), 0/5 (avg)   1         10        1          13s
```

modify the SO and delete the 2nd trigger:
```yaml
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: consumer
  namespace: keda2
spec:
  advanced:
    horizontalPodAutoscalerConfig:
      behavior:
        scaleDown:
          stabilizationWindowSeconds: 60
        scaleUp:
          stabilizationWindowSeconds: 60
  scaleTargetRef:
    name: consumer
  pollingInterval: 5
  minReplicaCount: 1
  maxReplicaCount: 10
  triggers:
  - type: rabbitmq
    metadata:
      protocol: amqp
      queueName: keda-tests
      queueLength: "5"
    authenticationRef:
      name: keda-trigger-auth-rabbitmq-conn
```

check the HPA:
```
kubectl get hpa keda-hpa-consumer
NAME                                  REFERENCE                               TARGETS                        MINPODS   MAXPODS   REPLICAS   AGE
keda-hpa-consumer                     Deployment/consumer                     0/5 (avg), <unknown>/5 (avg)   1         10        1          15m
```

notice that KEDA's metrics server is indeed not exposing metrics for the trigger that I deleted, hence the "unknown" current value, and can be confirmed in the events:
```
Events:
  Type     Reason                   Age                   From                       Message
  ----     ------                   ----                  ----                       -------
  Warning  FailedGetExternalMetric  24s (x26 over 6m47s)  horizontal-pod-autoscaler  unable to get external metric keda2/rabbitmq-keda-tests2/&LabelSelector{MatchLabels:map[string]string{scaledObjectName: consumer,},MatchExpressions:[]LabelSelectorRequirement{},}: unable to fetch metrics from external metrics API: No matching metrics found for rabbitmq-keda-tests2
```

### Expected Behavior

Keda operator should propagate all SO modifications to the underlying HPA

### Actual Behavior

The HPA stays with old triggers, which could cause unexpected behaviors.

### Steps to Reproduce the Problem

1. create a SO with multiple triggers
2. check SO & HPA resources
3. delete one of the SO triggers and apply 
4. check SO (doesn't show the deleted trigger) & HPA resources (still contains the deleted metric)

### Logs from KEDA operator

```
2021-03-25T08:44:34.107Z	INFO	controllers.ScaledObject	Creating a new HPA	{"ScaledObject.Namespace": "keda2", "ScaledObject.Name": "consumer", "HPA.Namespace": "keda2", "HPA.Name": "keda-hpa-consumer"}
2021-03-25T08:44:34.225Z	INFO	controllers.ScaledObject	Initializing Scaling logic according to ScaledObject Specification	{"ScaledObject.Namespace": "keda2", "ScaledObject.Name": "consumer"}
2021-03-25T08:44:34.298Z	INFO	controllers.ScaledObject	Reconciling ScaledObject	{"ScaledObject.Namespace": "keda2", "ScaledObject.Name": "consumer"}
2021-03-25T08:44:39.012Z	INFO	controllers.ScaledObject	Reconciling ScaledObject	{"ScaledObject.Namespace": "keda2", "ScaledObject.Name": "consumer"}
2021-03-25T08:54:54.408Z	INFO	controllers.ScaledObject	Reconciling ScaledObject	{"ScaledObject.Namespace": "keda2", "ScaledObject.Name": "consumer"}
2021-03-25T08:54:54.427Z	INFO	controllers.ScaledObject	Initializing Scaling logic according to ScaledObject Specification	{"ScaledObject.Namespace": "keda2", "ScaledObject.Name": "consumer"}
2021-03-25T08:55:06.532Z	INFO	controllers.ScaledObject	Reconciling ScaledObject	{"ScaledObject.Namespace": "keda2", "ScaledObject.Name": "consumer"}
```

### KEDA Version

2.2.0

### Kubernetes Version

1.18

### Platform

Any

### Scaler Details

RabbitMQ

### PR DESCRIPTION
**Pull request by coderanger**
Fixes #1699 and includes a test to make sure it doesn't come back.

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [x] Changelog has been updated

### CODE REVIEW

**Comment by coderanger** [comment]
This probably will fail in CI since I need to add stuff to the GHA config to install the envtest binaries for functional tests, but I will do that tomorrow. The fix itself can be cherrypicked if getting the envtest stuff in order ends up being a blocker.

**Comment by zroubalik** [comment]
> This probably will fail in CI since I need to add stuff to the GHA config to install the envtest binaries for functional tests, but I will do that tomorrow. The fix itself can be cherrypicked if getting the envtest stuff in order ends up being a blocker.

Thanks a lot for the fix! Add an entry to the changelog as well please.

File: controllers/hpa.go [hunk 1]
```diff
@@ -111,7 +111,8 @@ func (r *ScaledObjectReconciler) updateHPAIfNeeded(logger logr.Logger, scaledObj
 		return err
 	}
 
-	if !equality.Semantic.DeepDerivative(hpa.Spec, foundHpa.Spec) {
+	// DeepDerivative ignores extra entries in arrays which makes removing the last trigger not update things, so trigger and update any time the metrics count is different.
```
Comment (zroubalik) [inline controllers/hpa.go:114]: Wow 🤦

**Comment by coderanger** [comment]
@zroubalik I'm guessing I'll need to split that last commit into it's own PR, merge that, then rebuild this?

**Comment by zroubalik** [comment]
This PR needs to be rebased in order to get the new build tools image (the location was changed from `kedacore/build-tools` to `ghrc.io/kedacore/build-tools`) and use kubebuilder installed on that image.

**Comment by coderanger** [comment]
Looks like the tests are happy, so we should be all set to write more controller functional tests in the future as needed :)

**Review by zroubalik — APPROVED** [review]
Thanks a lot for this!

**Pull request merged**